src/node/rpc/frontend.h

```diff
@@ -464,11 +464,33 @@ namespace ccf
               auto tx_id = tx.get_txid();
               if (tx_id.has_value() && consensus != nullptr)
               {
-                // Only transactions that acquired one or more map handles
-                // have a TxID, while others (e.g. unauthenticated commands)
-                // don't. Also, only report a TxID if the consensus is set, as
-                // the consensus is required to verify that a TxID is valid.
-                ctx->set_tx_id(tx_id.value());
+                try
+                {
+                  endpoints.execute_endpoint_locally_committed(
+                    endpoint, args, tx_id.value());
+                }
+                catch (std::exception& e)
+                {
+                  // run default handler to set transaction id in header
+                  ccf::endpoints::default_locally_committed_func(
+                    tx_id.value(), args);
+                  ctx->set_error(
+                    HTTP_STATUS_INTERNAL_SERVER_ERROR,
```

Comment:
I've defaulted to a 500 for these errors as they are an internal error. I wonder if we could more subtly indicate to the user that something failed so that they can still see a success from the main logic and we just set the txid in the header. Maybe we could also set an error message in a header and otherwise leave the response alone?

e.g. `x-ms-ccf-locally-committed-error`: e.what()`

Reply:
Currently this would work since the locally_commited function can't get data from the main function without sending it through the response. However, with part 2 of the work we'll allow them to pass arbitrary data which may mean the response isn't filled at all, potentially emitting a far less useful body. Not sure if that would still be useful as we wouldn't be too intrusive.

Mainly from a client perspective I'd be interested in seeing the success response code to know the transaction succeeded even if I can't parse the body properly, rather than a complete failure message that actually did something.

Reply:
I think if the `committed_func` throws, we want to convert that into a HTTP error so its loud and obvious - ideally we'd enforce that at it doesn't throw at compile-time, but the next-best we can do is send a strong signal to the developer that it has thrown and they should fix that. Adding a header only in failing instances is likely to be missed (and potentially leave the rest of the response unparseable).

I see the appeal in trying to embed the original response inside this error, so a dedicated client _can_ still work out what happened, but we can't do that in general while maintaining the odata response body. I think this approach is fine - if it throws we discard the original response, and return a 500. Note that the app developer can still return something different if they wish - they can override `execute_endpoint_locally_committed`, have _that_ run a try-catch block, and return the original response/some encoding of it in the request, if they wish.